src/systemd-sonic-generator/ssg-test.cc

```diff
@@ -0,0 +1,550 @@
+
+/*
+ * This test class uses following directory hierarchy for input and output
+ * data for systemd-sonic-generator.
+ *
+ *    tests/ssg-test/   --- Test data directory
+ *             |
+ *             |---generated_services.conf
+ *             |---machine.conf (systemd-sonic-generator fetch platform from here)
+ *             |---systemd/
+ *             |       |--- *.service (Test unit files are copied from
+ *             |                       tests/testfiles/ to here)
+ *             |----test_platform/  (test platform)
+ *             |       |---asic.conf
+ *             |
+ *             |----generator/   (Output Directory)
+ *
+ */
+const std::string TEST_ROOT_DIR = "tests/ssg-test/";
+const std::string TEST_UNIT_FILE_PREFIX = TEST_ROOT_DIR + "systemd/";
+const std::string TEST_ASIC_CONF_FORMAT = TEST_ROOT_DIR + "%s/asic.conf";
+const std::string TEST_MACHINE_CONF = TEST_ROOT_DIR + "machine.conf";
+
+const std::string TEST_PLATFORM_DIR = TEST_ROOT_DIR + "test_platform/";
+const std::string TEST_ASIC_CONF = TEST_PLATFORM_DIR + "asic.conf";
+
+const std::string TEST_OUTPUT_DIR = TEST_ROOT_DIR + "generator/";
+
+const std::string TEST_CONFIG_FILE = TEST_ROOT_DIR + "generated_services.conf";
+
+const std::string TEST_UNIT_FILES = "tests/testfiles/";
+
+/* Input data for generated_services.conf */
+const std::vector<std::string> generated_services = {
+    "multi_inst_a.service",     /* Single instance of a  multi asic service a */
+    "multi_inst_a@.service",    /* Multi-instance of a multi asic service a */
+    "multi_inst_b@.service",    /* Multi-instance of a multi asic service b */
+    "single_inst.service",      /* A single instance service */
+    "test.service",             /* A single instance test service
+                                   to test dependency creation */
+    "test.timer",               /* A timer service */
+};
+
+static std::mutex g_ssg_test_mutex;
+
+class SystemdSonicGeneratorFixture: public testing::Test {
+  protected:
+    /* Save global variables before running tests */
+    virtual void SetUp() {
+        /* one test runs at a time */
+        g_ssg_test_mutex.lock();
+
+        unit_file_prefix_ = g_unit_file_prefix;
+        config_file_ = g_config_file;
+        machine_config_file_ = g_machine_config_file;
+        asic_conf_format_ = g_asic_conf_format;
+    }
+
+    /* Restore global vars */
+    virtual void TearDown() {
+        g_unit_file_prefix = unit_file_prefix_;
+        g_config_file = config_file_;
+        g_machine_config_file = machine_config_file_;
+        g_asic_conf_format = asic_conf_format_;
+
+        g_ssg_test_mutex.unlock();
+    }
+
+  private:
+    const char* unit_file_prefix_;
+    const char* config_file_;
+    const char* machine_config_file_;
+    const char* asic_conf_format_;
+};
+
+/* 
+ * class SsgFunctionTest
+ * Implements functions to execute functional level tests.
+ */
+class SsgFunctionTest: public SystemdSonicGeneratorFixture {
```

Comment:
Suggest to add a space before first `:` to align style with other SONiC C++ files.